apps/prairielearn/src/lib/server-jobs-2.ts

```diff
@@ -0,0 +1,184 @@
+import { z } from 'zod';
+import * as Sentry from '@prairielearn/sentry';
+import { logger } from '@prairielearn/logger';
+import { loadSqlEquiv, queryAsync, queryValidatedOneRow } from '@prairielearn/postgres';
+
+import { chalk } from './chalk';
+import serverJobs = require('./server-jobs');
+import socketServer = require('./socket-server');
+
+const sql = loadSqlEquiv(__filename);
+
+interface CreateServerJobOptions {
+  courseId?: string;
+  courseInstanceId?: string;
+  courseRequestId?: string;
+  assessmentId?: string;
+  userId?: string;
+  authnUserId?: string;
+  type: string;
+  description: string;
+}
+
+interface ServerJobExecOptions {
+  cwd?: string;
+  env?: NodeJS.ProcessEnv;
+}
+
+type ServerJobContext = Omit<ServerJob, 'execute' | 'executeInBackground'>;
+type JobExecutionFunction = (context: ServerJobContext) => Promise<void>;
+
+class ServerJob {
+  public jobSequenceId: string;
+  public jobId: string;
+  private started = false;
+  private finished = false;
+  public output = '';
+
+  constructor(jobSequenceId: string, jobId: string) {
+    this.jobSequenceId = jobSequenceId;
+    this.jobId = jobId;
+  }
+
+  error(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  warn(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  info(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  verbose(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  async exec(file: string, args: string[] = [], options: ServerJobExecOptions = {}): Promise<void> {
+    const msg = `$ ${file} ${args.join(' ')}\n`;
+    this.addToOutput(chalk.greenBright(msg));
+    const proc2 = execa(file, args, {
+      ...options,
+      reject: false,
+      all: true,
+    });
+
+    proc2.all?.setEncoding('utf-8');
+    proc2.all?.on('data', (data) => {
+      this.addToOutput(data);
+    });
+
+    const result = await proc2;
+    if (result.exitCode !== 0) {
+      throw new Error(`Process exited with code ${result.exitCode}`);
+    }
+  }
+
+  /**
+   * Runs the job sequence and returns a Promise that resolves when the job
+   * sequence has completed. The returned promise will not reject if the job
+   * sequence fails.
+   */
+  async execute(fn: JobExecutionFunction): Promise<void> {
+    if (this.started) {
+      throw new Error('ServerJob already started');
+    }
+    this.started = true;
+
+    try {
+      await fn(this);
+      await this.finish();
+    } catch (err) {
+      try {
+        await this.finish(err);
+      } catch (err) {
+        logger.error(`Error failing job ${this.jobId}`, err);
+        Sentry.captureException(err);
+      }
+    }
+  }
+
+  /**
+   * Identical to {@link execute}, except it doesn't return a Promise. This is
+   * just used as a hint to the caller that they don't need to wait for the job
+   * to finish. Useful for tools like TypeScript and ESLint that ensure that
+   * promises are awaited.
+   */
+  executeInBackground(fn: JobExecutionFunction): void {
+    this.execute(fn);
```

Comment:
Maybe we should do the `this.started` check here too? Otherwise we'd throw an error that would never be caught since nothing awaits this promise.

Reply:
I'm in favor of anything that can make sure we never throw errors that will be lost.

Reply:
Changed in 90f94d6.